devito/mpi/halo_scheme.py

```diff
@@ -28,14 +29,41 @@ class HaloLabel(Tag):
 STENCIL = HaloLabel('stencil')
 
 
-HaloSchemeEntry = namedtuple('HaloSchemeEntry', 'loc_indices loc_dirs halos dims')
+class HaloSchemeEntry(Reconstructable):
+
+    __rargs__ = ('loc_indices', 'loc_dirs', 'halos', 'dims')
+
+    def __init__(self, loc_indices, loc_dirs, halos, dims):
+        self.loc_indices = frozendict(loc_indices)
+        self.loc_dirs = frozendict(loc_dirs)
+        self.halos = halos
+        self.dims = dims
+
+    def __eq__(self, other):
+        if not isinstance(other, HaloSchemeEntry):
+            return False
+        return (self.loc_indices == other.loc_indices and
+                self.loc_dirs == other.loc_dirs and
+                self.halos == other.halos and
+                self.dims == other.dims)
+
+    def __hash__(self):
```

Comment:
why do you need to use all these frozentsets?

if `halos` and `dims` are mutuable, they should be turned into immutable inside `__init__`

Reply:
is this fine now, or I still do not understand?